source/common/common/bit_array.h

```diff
@@ -0,0 +1,152 @@
+ *  3) It works with a maximum width of 32-bits
+ */
+class BitArray {
+public:
+  /**
+   * Constructs a BitArray.
+   * @param width the fixed width of the array elements. For example, if we need
+   *  to represent 10 distinct values we need at least 4 bits as the width.
+   * @param num_items the number of elements the bit array must hold.
+   */
+  BitArray(int width, size_t num_items)
+      : array_start_(std::make_unique<uint8_t[]>(bytesNeeded(width, num_items))), bit_width_(width),
+        // This will fit in a uint32_t as with the maximum shift of 32, we'd
+        // subtract one to fit in 32 bits.
+        mask_(static_cast<uint32_t>((static_cast<uint64_t>(1) << width) - 1)),
+        num_items_(num_items) {
+    RELEASE_ASSERT(width <= MaxBitWidth, "Using BitArray with invalid parameters.");
+    static_assert(
+        sizeof(uint8_t*) == 8,
+        "Pointer underlying size is not 8 bytes, are we running on a 64-bit architecture?");
+    // Init padding to avoid sanitizer complaints if reading the last elements.
+    uint8_t* padding_start = array_start_.get() + (bytesNeeded(width, num_items) - WordSize);
+    storeUnsignedWord(padding_start, 0);
+  }
+
+  static constexpr int MaxBitWidth = 32;
+  static constexpr int WordSize = sizeof(uint8_t*);
+
+  /**
+   * Gets the fixed width bit at the given index.
+   * Assumes the the provided index is in bounds.
+   */
+  inline uint32_t get(size_t index) const {
+    RELEASE_ASSERT(index < num_items_, "BitArray requested index out of bounds");
+    // We locate the first byte that contains part of the element located at
+    // the given index.
+    const size_t bit0_offset = index * bit_width_;
+    const size_t byte0_offset = bit0_offset >> 3;
+    const uint8_t* byte0 = array_start_.get() + byte0_offset;
+    // Find the starting bit within byte0 of this element.
+    // It will be in the range of 0-7.
+    const size_t index_of_0th_bit = bit0_offset & 0x7;
+    // We then load regardless of alignment an entire word to get the bytes
+    // containing the element. We shift this to get have the element's bits
+    // at bit 0, and then apply the mask which will capture all the bits
+    // pertaining to this element.
+    return (LoadUnsignedWord(byte0) >> index_of_0th_bit) & mask_;
+  }
+
+  /**
+   * Sets the given fixed width bit to the given value.
+   */
+  inline void set(size_t index, uint32_t value) {
+    if (index > num_items_) {
+      ENVOY_BUG(true, "BitArray::set requested index out of bounds");
+      return;
+    }
+    // We locate the first byte that contains part of the element located at
+    // the given index.
+    const size_t bit0_offset = index * bit_width_;
+    const size_t byte0_offset = bit0_offset >> 3;
+    uint8_t* byte0 = array_start_.get() + byte0_offset;
+
+    // Find the starting bit within byte0 of this element.
+    // It will be in the range of 0-7.
+    const size_t index_of_0th_bit = bit0_offset & 0x7;
```

Comment:
nit: you could probably get away (performance wise) with factoring out the common parts of this with get() into common helper function(s). It should all be inlined by the compiler anyway.

Reply:
Leaning towards keeping the code where it is so it can flow better with the comments, context.

Reply:
ok, sgtm